Add BaseDataset and FileSystemDataset classes

convexquad opened the pull request:
Incorporate the changes from the reviews for https://github.com/woven-planet/wicker/pull/57 in this PR:
- We leave the `AbstractDataset` class alone so that it remains a pure interface.
- As suggested in the previous review, we create the `BaseDataset` class as a concrete implementation of `AbstractDataset` that is orthogonal to any filesystem or bucket-specific assumptions (it will rely on the `storage` and column file reader instances for this logic).

The `S3Dataset` class is updated so that it is a subclass of `BaseDataset` (there is still a little bit of S3-specific logic that is left in this class for backwards compatibility).

We create the new `FileSystemDataset` class, but it only exists so that its constructor arguments can be constrained to be compatible with local storage (e.g. the `storage` argument must have the type `FileSystemDataStorage` and the Arrow filesystem has to be a `LocalFileSystem`). However, I would be ok to remove this class if the reviewers feel it is unnecessary.

Testing:
- Unit tests added to cover the new `FileSystemDataset` class and its new behavior to support local storage (together with `BaseDataset`).
- Test training jobs: I cannot add a link to the training job since this is a public GitHub repository, but I have tested this PR in our ML repository on my branch `dev/abain/test_wicker_s3` with run ID `9eb6efad-6c64-4260-abec-a5d3c0c59779`. The test training job loads many `S3Dataset` instances. The test training job completed successfully.

I have been able to make this PR is totally orthogonal from https://github.com/woven-planet/wicker/pull/73 . Although I think #73 has good changes for Wicker, actually this PR is the one we really need to enable reading Wicker datasets from local filesystems.

Note: One thing Zhenyu proposed in his review of Isaak's PR is to change the user experience so that they use a factory to return instances of `AbstractDataset`. We want to leave this new user experience up to the platform team. However, the new factory methods should work well with the changes in this PR!

Commit:
Add BaseDataset and FileSystemDataset classes

convexquad commented on tests/test_storage.py:
```diff
@@ -37,7 +37,7 @@ def test_fetch_file(self) -> None:
             # create local file store
             local_datastore = FileSystemDataStorage()
             # save file to destination
```
The change below just fixes a bug in the test.

convexquad commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
         pass
 
```
Please pardon, I must be super OCD and re-order the `arrow_table` function to come before `schema` function so that they are in alphabetic order. Maybe I need to go to coding mental hospital.

zhenyu replied:
Cool， I am good with reformat at once

convexquad commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        """Return the schema of the dataset."""
         pass
 
 
```
For the `BaseDataset` class, some of the key arguments like `pa_filesystem`, `path_factory` and `storage` should no longer be `Optional` or take an S3-specific default value. So in the `__init__` function, I just move these arguments higher in the argument list. Pardon that GitHub has a little trouble and makes the diff look more complicated than this.

convexquad commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
-            pafs.S3FileSystem(region=get_config().aws_s3_config.region) if pa_filesystem is None else pa_filesystem
+        # Create the column bytes reader that depends on the other init parameter values.
+        self._column_bytes_file_reader = self._build_column_bytes_reader()
+
```
In the rest of the `BaseDataset` class, we can move most of the functions like `__getitem__` and `__len__` from `S3Dataset` into this class with no changes. The only thing we need to add to this class is a function to determine what kind of column bytes reader to build.

convexquad commented on wicker/core/datasets.py:
```diff
@@ -207,6 +272,104 @@ def schema(self) -> DatasetSchema:
                 )
         return self._schema
 
+
```
As I mentioned in the PR description, now the only thing that `FileSystemDataset` has is an `__init__` function that further constrains the types of its arguments like `storage` and `pa_filesystem`. This can help the user to reduce any type-related errors. However, I would be ok to remove this class if the reviewers do not think it is very useful.

convexquad replied:
Another way to do things would be to remove `pa_filesystem`, `path_factory` and `storage` as arguments to the `__init__` function and instead add factory methods to the class that build an appropriate instance for each of these. This enables them to be removed from the constructor and also enables subclasses to override the factory methods in case they want to pass a more specific subtype.

zhenyu replied:
From the Ioc and testable class design. I like your implementation, I would suggest a get_filesystemdataset factory to decouple the construction detail and user interface as I suggested in the previous FileSystemDataset PR by @pickles-bread-and-butter

Commit:
Fix tests

Commit:
Fix tests

Commit:
Type fix

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
 
-class S3Dataset(AbstractDataset):
-    """Implementation for a Map-based dataset"""
+class BaseDataset(AbstractDataset):
```
Do we want to use inherit  or combination to implement the Abstract interface. To me, it seems we lean to the combination style, since we inject the dependency. If so, prob BaseDataset->DatasetImpl and prob the Abstract interface is not necessary, the external customer would not expect to use the arrow directly. WDYT?

convexquad replied:
Let me mark as resolved since we are adding a factory class that users can use to create instances instead, so we may not have to worry so much about the `BaseDataset` class. But please feel free to reopen and let me know your thoughts about any changes we should make to this class.

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
         dataset_version: str,
         dataset_partition_name: str,
-        local_cache_path_prefix: str = os.getenv("TMPDIR", "/tmp"),
+        pa_filesystem: pafs.FileSystem,
```
Curious with the relationship of this to the DataStorage class, seems like concept overlap

convexquad replied:
Yes, of course the storage type should determine the arrow file system type. Maybe sometime over good coffee I can explain the history with my partner on the project. He was smart and fast programmer (now he is cofounder of the company "Daft Dataframe"), but at least then he was super wild. I did not get to review this code.

But back to the issue. Since the current `S3Dataset` [constructor](https://github.com/woven-planet/wicker/blob/main/wicker/core/datasets.py#L150) accepts a `pa_filesystem` object, I'm not sure if we can fix the issue or not since we need to keep backwards compatibility with the current code.

Here in `BaseDataset` constructor, we could remove `pa_filesystem` from the `__init__` parameters and set it as a member variable by calling a new function `self._pa_filesystem = self._storage.arrow_file_system()` that we add to the `AbstractStorage` class. Then to maintain backwards compatibility in the `S3Dataset` constructor, we can just leave the current logic [here](https://github.com/woven-planet/wicker/blob/main/wicker/core/datasets.py#L186) and override the `self._pa_filesystem` member variable. That is my best idea - let know what you think!

zhenyu replied:
Thanks for the explanation, I know this is not introduced by your, but from the concept, the original wicker author reinvent the datastorage class but hard coded the abstract with arrow lib. Really did not get their point of the design, make me really confusion.
No matter what they are doing, at least at that time, not good programmer and did not fix the tech debt

zhenyu replied:
We can do this refactor after this PR

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        """Return the schema of the dataset."""
         if self._schema is None:
-            schema_path = self._s3_path_factory.get_dataset_schema_path(self._dataset_id)
+            schema_path = self._path_factory._get_dataset_schema_path(self._dataset_id)
```
Is the schema_path abs S3 path? I did not see a local file path factory, correct me if I was wrong

zhenyu replied:
My point is per my understanding the path should be different between the mounted filesystem and S3 path, correct me if I was wrong.

convexquad replied:
For `FileSystemDataset` we work with the `WickerPathFactory` base class that is basically like local path factory (`S3PathFactory` is a subclass of `WickerPathFactory`).

So in this line, `schema_path` will end up being an absolute local path for `FileSystemDataset` and it will be an absolute S3 path for `S3Dataset`.

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        """
+        return len(self.arrow_table())
+
+    def _build_column_bytes_reader(self) -> ColumnBytesFileReader:
```
should we make it a singleton?

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
+        """Builder function to return a column bytes file reader."""
+        if self._local_cache_path_prefix:
+            # If caching column files from storage into the local filesystem, return a cache-enabled reader.
+            return ColumnBytesFileCache(
```
I would like to add a columfileReader para to the __init__ of the basedataset and push down this cache vs no-cache builder to the subclass. 
in the s3 dataset build the ColumnBytesFileCache add inject it the BaseDataset
in the filesystem dataset build either ColumnBytesFileCache or ColumnBytesFileReader based on the FuseFilesystem cache cap. By this mean, I would high suggest a get_dataset factory or builder method. For the new introduced FileSystemDataset, we should hide the detail to improve later. Otherwise, we just mix the Interface and Implementation together like the original Wicker S3Dataset, by that way. No point for an abstract dataset

convexquad replied:
This make sense; I added the column bytes file reader parameter to the `BaseDataset.__init__` function and pushed down its construction to the derived classes as suggested above. The `S3Dataset` class will always build a a `ColumnBytesFileCache` instance. I think it's fine for now if the `FileSystemDataset` always builds a `ColumnBytesFileReader` instance.

Commit:
Add column_bytes_file_reader parameter to BaseDataset init

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -207,6 +255,119 @@ def schema(self) -> DatasetSchema:
         return self._schema
 
+
+class FileSystemDataset(BaseDataset):
```
Make a factory method or builder to avoid repeating the S3Dataset mistake again.

convexquad replied:
@zhenyu one second I'm not finished!

convexquad replied:
@zhenyu I just added it. Could you review the new `DatasetFactory` class? I am happy to make adjustments. I added factory methods for both `create_filesystem_dataset` and `create_s3_dataset`, but we could remove the S3 method if you don't want it. For `create_filesystem_dataset` we could make more adjustments to its parameters and/or the `FileSystemDataset.__init__` function parameters.

convexquad replied:
Since the default for `FileSystemDataset` should be no-cache, let me suggest that a very slight alternative plan!

Let's leave the name of the current class as `FileSystemDataset`. But if we ever have a version or subclass of `FileSystemDataset` that caches files (by copying them from the mounted FUSE root directory into another directory on the instance's local disk), let's give that one a more specific name (like `CacheFileSystemDataset` or `FileSystemCacheDataset`).

Commit:
Add DatasetFactory class to hide implementation details

Commit:
Fix test

convexquad commented:
Thank you @zhenyu for the review so far! Since we have made some changes together, let me get another review when you can (especially for the new factory class and functions)!

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -303,3 +452,85 @@ def dataset_size(self) -> int:
         return self._get_dataset_partition_size()
+
+
+class DatasetFactory:
```
Thanks a lot for adding this, I think a function is enough, we do not sub class the factory since the dependency injection is not so complex. 
I am fine with a factory class if you think we will sub class eventually

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -303,3 +452,85 @@ def dataset_size(self) -> int:
+    Wicker dataset that you want.
+    """
+
+    def create_filesystem_dataset(
```
Can we just expose a create dataset method and return the abstract dataset interface?

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -303,3 +452,85 @@ def dataset_size(self) -> int:
+        dataset_name: str,
+        dataset_version: str,
+        dataset_partition_name: str,
+        root_path: str,
```
Can this root_path config in the Wicker config json?

convexquad replied:
Yes, I updated the `wickerconfig.json` file format to include a `filesystem_config` entry that includes its root path.

Commit:
Replace factory class with builder function

zhenyu commented on wicker/core/config.py:
```diff
@@ -40,6 +40,7 @@ def from_json(cls, data: Dict[str, Any]) -> BotoS3Config:
 
 @dataclasses.dataclass(frozen=True)
 class WickerAwsS3Config:
+    loaded: bool = False
```
I know it is not introduced by you, but I think the WickerAwsS3Config should be singleton instead of class.

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -303,3 +452,58 @@ def dataset_size(self) -> int:
+    .. seealso:: `filters in <https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html>`__ # noqa
+    """
+    config = get_config()
+    if config.filesystem_config.loaded:
```
I think the type of the dataset should be one of the input parameter, instead of controlled by config. There maybe use case that need to create both S3datasets and FilesystemDatasets at the same time

convexquad replied:
@zhenyu I was thinking the same thing. Then I started to think we might need to have multiple `WickerAwsS3Config` or `WickerFileSystemConfig` entries in the `wickerconfig.json` configuration file.

Do you think that is a possible usecase? If so, then I can add a `configuration_name` property to `WickerAwsS3Config` and `WickerFileSystemConfig` and change our new builder function so that it accepts a particular `configuration_name`.

Hower, if `wickerconfig.json` can only contain exactly one `WickerAwsS3Config` and `WickerFileSystemConfig`, then I will change the builder function to accept the type of the requested dataset.

zhenyu replied:
Do you think that is a possible usecase?  I think at least we should not block the customer to do some possible options.
Second, I think we should have multiple entries in the config file for each Naming space. We only need one S3 because s3 is a naming space, we prob need more filesystem config because each mounted Dir is a naming space. WDYT? 
I think the original wicker mix the naming space with the storage class and filesystem concept together. Let us improve it little by little

convexquad replied:
@zhenyu Yes, I am thinking one S3Dataset but multiple FileSystemDatasets is a usecase. I will push a short update to support this.

Commit:
Fix builder function to accept a dataset configuration type and update tests

Commit:
Fix type-check

convexquad commented:
@zhenyu I fixed our new builder function to accept a `dataset_config` parameter to enable the user to easily specify the dataset type (and I updated the `wickerconfig.json` file to support `FileSystemDataset` configuration). Everything looks good. But now as you can see from my last comment I am wondering if the `wickerconfig.json` file should support multiple `WickerAwsS3Config` or `WickerFileSystemConfig` named entries, do you think this is a usecase or I am over-thinking it.

Commit:
Support config for one S3Dataset and multiple FileSystemDatasets

convexquad commented:
@zhenyu I think I have got it. Now you can have `wickerconfig.json` like this:
```
{
  "aws_s3_config": {
    "s3_datasets_path": "s3://fake_data/",
    "region": "us-west-2",
    "boto_config": {
      "max_pool_connections":10,
      "read_timeout_s": 140,
      "connect_timeout_s": 140
    }
  },
  "filesystem_configs": [
    {
      "config_name": "filesystem_1",
      "root_datasets_path": "/mnt/bucket_1/"
    },
    {
      "config_name": "filesystem_2",
      "root_datasets_path": "/mnt/bucket_2/"
    }
  ],
  ... (other stuff)
}
```
There is still just the one `aws_s3_config`, but now you can have multiple filesystem configs for mounting multiple `FileSystemDataset` volumes into the training job. In the config, you have to give them a `configuration_name` property so that in the builder function you can know for which one you want to return a `FileSystemDataset` instance.

zhenyu commented on wicker/core/datasets.py:
```diff
@@ -303,3 +455,76 @@ def dataset_size(self) -> int:
+    return None
+
+
+def build_dataset(
```
👍 . Minor thing lean to you is from the testable/readable perspective, I would think make 2 private sub function like _build_s3_dataset _build_filesystem_dataset. 
Anyway, much better than the original S3Dataset interface. Thanks

zhenyu commented:
@convexquad Thanks so much for this PR. The class design is much better now although we could improve even further, let us do it little by little. For this PR, I am fine with the class design now. Please make sure enough test, the CI pipeline is not so convincing now. 
Ping me for an approval, once you think the test enough. Thanks again

convexquad commented:
@zhenyu thanks for all the reviews! I am running test cloud jobs to make sure everything is good with this. I might leave this PR open for a short while.

convexquad commented:
@zhenyu could you give me an approving review for this PR? I have completed test training jobs with both FUSE-mounted local `FileSystemDataset` instances as well as `S3Dataset` instances from this branch.

I cannot add a link to the training job since this is a public GitHub repository, but I have tested this PR in our ML repository on my branch `dev/abain/test_wicker_s3` with run ID `9eb6efad-6c64-4260-abec-a5d3c0c59779`. The test training job loads many `S3Dataset` instances.

zhenyu reviewed (approved):
Thanks @convexquad for this PR

marccarre reviewed (approved):
Thanks a lot for this PR @convexquad! 🙇🏻‍♂️ 

> _Running test training jobs is in progress; they will be linked here._

What is the status on this front? 👀

Commit:
Version bump
